test/e2e/user_certs_test.go

```diff
@@ -0,0 +1,281 @@
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/cert"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	operatorclient "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
+
+	test "github.com/openshift/cluster-kube-apiserver-operator/test/library"
+)
+
+func TestNamedCertificates(t *testing.T) {
+
+	namedServingCertsToCreate := map[string][]string{
+		"one":   {"one.test"},
+		"two":   {"two.test"},
+		"three": {"three.test", "four.test"},
+	}
+
+	kubeConfig, err := test.NewClientConfigForTest()
+	require.NoError(t, err)
+	kubeClient, err := clientcorev1.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+	configClient, err := configclient.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+	operatorClient, err := operatorclient.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+
+	// kube-apiserver must be available, not progressing, and not failing to continue
+	test.WaitForKubeAPIServerClusterOperatorAvailableNotProgressingNotFailing(t, configClient)
+
+	// create a root certificate authority crypto materials
+	rootCA := test.NewCertificateAuthorityCertificate(t, nil)
+
+	// create named serving certificate crypto materials
+	servers := createNamedServingCerts(t, namedServingCertsToCreate, rootCA.Certificate)
+
+	// create secrets for named serving certificates
+	secretNames := map[string]string{}
+	for id, server := range servers {
+		name := strings.ToLower(test.GenerateNameForTest(t, "Server-"+id+"-"))
+		defer func() {
+			err := deleteSecret(kubeClient, "openshift-config", name)
+			require.NoError(t, err)
+		}()
+		_, err := createTLSSecret(kubeClient, "openshift-config", name, server.PrivateKey, server.Certificate)
+		require.NoError(t, err)
+		secretNames[id] = name
+	}
+
+	// before updating the config, note current generation of KubeApiServer/cluster.
+	initialConfigGeneration := test.GetKubeAPIServerOperatorConfigGeneration(t, operatorClient)
+
+	// configure named certificates
+	defer func() {
+		_, err := updateAPIServerClusterConfigSpec(configClient, &configv1.APIServerSpec{})
+		require.NoError(t, err)
+	}()
+	_, err = updateAPIServerClusterConfigSpec(configClient, &configv1.APIServerSpec{
+		ServingCerts: configv1.APIServerServingCerts{
+			NamedCertificates: []configv1.APIServerNamedServingCert{
+				{ServingCertificate: configv1.SecretNameReference{Name: secretNames["one"]}},
+				{ServingCertificate: configv1.SecretNameReference{Name: secretNames["two"]}},
+				{ServingCertificate: configv1.SecretNameReference{Name: secretNames["three"]}},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// wait for configuration to become effective
+	test.WaitForNextKubeAPIServerOperatorConfigGenerationToFinishProgressing(t, operatorClient, initialConfigGeneration)
```

Comment:
is this really stable? There could be other parallel changes. Do we have a way to really check that the config was rolled out? we could check the last apiserver configmap to look for one of our cert name, maybe make it random so it is unique.